PWM testing/fixes

pkendall64 opened the pull request:
This PR is for people that wish to test the fixes for EESP8285 based PWM receiver reboot issue. #2577

There are a couple of changes here...
1. revert the fix for 0/100% 10kHz as the second commit does it simpler, outside the NMI handler.
2. simpler fix for 0/100% 10kHz
3. Optimise some of the NMI hander code
4. Extend the 4uS minimum interrupt to 10uS (this alone fixed the issue for me on 3.3.2)

The final commit is to disable both the software and hardware watchdog timers so if the issue does present itself it will hard fail rather than reboot. This is temporary and only for testing this PR.

I am currently running a long test to ensure this problem is resolved. Currently at 48hrs at the time of opening this PR.
I will leave this running for at least 7 days, if it is still running after that time, I'll call it good.

As mentioned below a hi packet rate (F1000) and a high PWM update rate (333/400Hz) exacerbates the issue.

JyeSmith commented:
Currently testing with 6 different receivers, and so far have a combined total of 40hrs up time without issue.

For anyone else testing please use F1000 and a PWM rate of 400Hz.

![image](https://github.com/ExpressLRS/ExpressLRS/assets/14170229/f4bb12eb-bb22-4b37-a996-29f276ffa378)

pkendall64 commented:
For anyone thats following...
First receiver running for 74.5hrs + 3 others at 24 hours = 146hours of continuous connection with no hang... testing will continue 😃

pkendall64 commented:
Still following along?
First receiver running for 127 hours + 3 others at 76.5 hours = 356.5 hours of continuous connection with no hang... testing will continue 😃

pkendall64 commented:
I'm calling it good.
First receiver running for 168 hours + 3 others at 117.5 hours = 520.5 hours of continuous connection with no hang.

wvarty commented on src/lib/PWM/waveform_8266.cpp:
```diff
@@ -257,33 +257,30 @@ static IRAM_ATTR void timer1Interrupt() {
           wvfState.waveformState ^= mask;
-          if (wvfState.waveformState & mask && wave->nextHighLowUs != 0) {
+          if (wvfState.waveformState & mask) {
+            if (i == 16) {
```
What is 16?
Magic?

pkendall64 replied:
It's the pin number GPIO16. Theres no constant we can use for that. This comes from the original waveform code from the Arduino framework code.

wvarty replied:
Can't add a `#define GPIO16 16` above it? With a comment?

wvarty reviewed (approved):
Feel free to ignore the magic 16 comment if you want. Its minor.

pkendall64 commented:
> Feel free to ignore the magic 16 comment if you want. Its minor.

I renamed the `i` and `pin` variables to `gpio` as thats actually what thre are, and added a comment on the line. Should be more obvious now.

CapnBry commented:
haha in typical me fashion, I spent the weekend testing this PR. I was unable to get my test RX (ER5Cv2) to lock up at all on either master or this PR. I ran F1000 and both 333Hz and 50Hz PWM channels (simultaneously). I mostly tested using a switch as input so it makes sense I didn't see lockups on f6de862bb230169ef00444b1102f006412c5f238 (master before this merge) due to the value not changing.

On this PR though, I see increased jitter in the servo output:

## Before
Note that the persistence bar of the servo output shows about 8us of jitter in the output signal

### F1000 333Hz 
![DS1Z_QuickPrint29](https://github.com/ExpressLRS/ExpressLRS/assets/677183/f17c98ed-5193-49b7-bb8d-cec16cdca5db)

### F1000 50Hz
![DS1Z_QuickPrint30](https://github.com/ExpressLRS/ExpressLRS/assets/677183/bea55f55-01d9-4725-86ef-f49a68f53b4a)

## PR

Now note that there's excursions from the regular 8us jitter that sometimes precede the expected trailing edge, and there are updates that happen around once every minute where the trailing edge has 2-3x as much jitter as before this.

### F1000 333Hz 
![DS1Z_QuickPrint31](https://github.com/ExpressLRS/ExpressLRS/assets/677183/423463af-b699-4588-87c1-4498fd4355a1)

### F1000 50Hz
![DS1Z_QuickPrint32](https://github.com/ExpressLRS/ExpressLRS/assets/677183/a8484387-6348-41ff-8c0b-6f3310bda152)

I've also tried 100Hz FullRes to see if the lower update rate removed the extra jitter but that does not seem to be the case. My _perception_ is that they happen more frequently on F1000 but not significantly. I also have run the code from before this PR for 2 hours and gotten zero excursions compared to quite frequently with this PR.

pkendall64 commented:
@CapnBry the extra jitter is probably because of this change

![Screenshot 2024-03-26 at 3 44 39 PM](https://github.com/ExpressLRS/ExpressLRS/assets/512740/06cefd9e-3c1d-4c67-9e2d-f7d2c26434e7)

This is the change that made the 3.3.2 version not crash.
